### PR TITLE
edit command breaks for EDITOR path containing spaces

### PR DESCRIPTION
**Pull request by omus**
I noticed when using the `edit` function in Julia that I was getting this message:

``` julia
julia> edit(edit, (String,))
Unknown editor: no line number information passed.
The method is defined at line 56.
```

After doing some digging I noticed that my default editor Sublime is supported by the `edit` function but something else was going on. My editor is set using the environmental variable EDITOR and looks like this:

``` bash
$ echo $EDITOR
/Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl
```

In Julia the path includes the backslash but is not recognized as a path:

``` julia
julia> ENV["EDITOR"]
"/Applications/Sublime\\ Text.app/Contents/SharedSupport/bin/subl"

julia> ispath(ENV["EDITOR"])
false
```

Since the `ispath` function doesn't recognize this string as a path the `edit` function will not recognize the editor as "subl" and state that line number support is not available. Note that if I set the EDITOR to not include the backslash `ispath` will pass but will cause the spawn call to fail:

``` julia
julia> ENV["EDITOR"] = "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"
"/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"

julia> ispath(ENV["EDITOR"])
true

julia> edit(edit, (String,))
ERROR: could not spawn `/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl /usr/local/Cellar/julia/HEAD/bin/../share/julia/base/interactiveutil.jl:56`: no such file or directory (ENOENT)
 in _jl_spawn at process.jl:255
 in anonymous at process.jl:408
 in setup_stdio at process.jl:396
 in spawn at process.jl:407
 in edit at interactiveutil.jl:37
 in edit at interactiveutil.jl:58
```

I have updated the code to use `shell_split` earlier which strips the backslash and makes it easier to find the editor name. Another improvement is allowing the if the user includes flags in their EDITOR (such as ".../subl -w") we can still determine the editor name.

I believe that this pull request fixes the issue in #11305 and is an improvement over the pull request #11329.

### CODE REVIEW

**Comment by stevengj** [comment]
I'm confused, shouldn't your `EDITOR` environment variable be `"/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"`?   Why does it include an explicit backslash character?

**Comment by omus** [comment]
@stevengj After further inspection I believe you are correct and I shouldn't be including the backslash. For some context here is what I had for my Sublime terminal setup:

```bash
SUBL2="/Applications/Sublime\ Text\ 2.app/Contents/SharedSupport/bin/subl"
SUBL3="/Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl"
SUBL=$SUBL3

alias subl2="$SUBL2"
alias subl3="$SUBL3"
alias subl="$SUBL"

export EDITOR="$SUBL"
```

The reason I included the backslash was to make the aliases work. Without the backslash I was getting `-bash: /Applications/Sublime: No such file or directory`. Upon further inspection I found that `crontab -e` was broken using this setup. After some tinkering my correct terminal setup is now:

```bash
SUBL2="/Applications/Sublime Text 2.app/Contents/SharedSupport/bin/subl"
SUBL3="/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"
SUBL=$SUBL3

# Note: Single quotes used to handle spaces
alias subl2="'$SUBL2'"
alias subl3="'$SUBL3'"
alias subl="'$SUBL'"

export EDITOR="$SUBL"
```

Note that my updated setup still has problems with Julia's `edit` method not working correctly. I'll do some additional work and see if I can find a better solution.

**Comment by omus** [comment]
Now after running several experiments the EDITOR variable using `crontab -e`, `git commit`, and Julia's `edit` method I think we definitely should be escaping an EDITOR path that contains spaces.

My rational behind this decision is that the EDITOR variable can contain more than just a path to an executable and can contain flags (note I can't find any official sources on this). When an EDITOR contains both spaces in the executable path and flags it becomes difficult to determine the editor's name. 

eg.
```julia
julia> shell_split("/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl -w")
3-element Array{AbstractString,1}:
 "/Applications/Sublime"                   
 "Text.app/Contents/SharedSupport/bin/subl"
 "-w"
```
vs.
```julia
julia> shell_split("/Applications/Sublime\\ Text.app/Contents/SharedSupport/bin/subl -w")
2-element Array{AbstractString,1}:
 "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"
 "-w" 

julia> shell_split("'/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl' -w")
2-element Array{AbstractString,1}:
 "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"
 "-w"                                                            
```

Alternatively, if we really wanted to we could support an EDITOR without escaped spaces by assuming that no flags will exist. Note that the command `git commit` works with flags and requires escaped spaces where as `crontab -e` only works when variable is an unescaped string without flags.

One last note: I'll be adding the following to my terminal setup to keep crontab and Julia happy:

```julia
export JULIA_EDITOR="'/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl'"
```

**Comment by omus** [comment]
@stevengj Here is the short answer to your original question:

The backslash was included to escape the path such that the `spawn` call could find the executable. Without it `spawn` tries to run `/Applications/Sublime` which isn't a valid executable.

eg.
```julia
julia> ENV["EDITOR"] = "/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"
"/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl"

julia> edit(edit, (String,))
ERROR: could not spawn `/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl /usr/local/Cellar/julia/HEAD/bin/../share/julia/base/interactiveutil.jl:56`: no such file or directory (ENOENT)
 in _jl_spawn at process.jl:255
 in anonymous at process.jl:408
 in setup_stdio at process.jl:396
 in spawn at process.jl:407
 in edit at interactiveutil.jl:37
 in edit at interactiveutil.jl:58
```

**Comment by stevengj** [comment]
@omus, it is a bug in `jl_spawn` if it can't handle executable paths with spaces.

See also [here](http://superuser.com/questions/616351/spaces-in-linux-environmental-variables) ... if you have paths with spaces, you should generally *not* escape them in environment variables, but rather you should properly quote them when they are used.

**Comment by stevengj** [comment]
In particular, it looks like the culprit is the [call to shell_split](https://github.com/JuliaLang/julia/blob/2232315e2093e30f519be30eae5e0004e005dc42/base/interactiveutil.jl#L37) in the `edit` function, not in `spawn`.

The bug was introduced by #6865 by @jayschwa in order to handle sublime arguments, but the "cure" introduced a bug whenever the sublime path contains a space.

**Comment by stevengj** [comment]
In general, any call to `shell_split` seems to have this problem.   I'm not sure what the right solution is.  How do we distinguish between executing the file named "foo file" and executing the file "foo" with the argument "file"?

I guess we could call `shell_split` only starting at hyphen preceded by whitespace, although that heuristic would fail for a file perversely named "foo -bar" (which is a perfectly valid filename).

**Comment by stevengj** [comment]
(But in any case, the proposed solution, of recognizing backslashed spaces in a pathname, is definitely wrong.)

**Comment by omus** [comment]
@stevengj I'll note the current implementation of `edit` (prior to this pull request) does work with an escaped pathname using a backslash. My fix just lets `edit` determine the editor name of the path in question.

**Comment by omus** [comment]
I agree that escaping/quoting should happen as late as possible. In this case however I don't think you can escape later due to the example you illustrated with `foo file`. What if you have `foo bar baz`? Is this the executable "foo" with two arguments? Maybe it is `'foo bar' baz` or `'foo bar baz'`.

**Comment by stevengj** [comment]
@omus, the problem is that there are two conflicting desires here: 

* support paths with spaces: favors `spawn(`$editor $file -l $line`)`
* support `EDITOR` commands that include options (`sublime --wait`): requires you to split `editor` into multiple strings passed to `spawn`: use `shell_split`

Your patch fixes the former at the price of breaking the latter (#6865).  Maybe this is a good idea, or maybe there is a heuristic to mostly support both (e.g. split after substrings beginning with `" --"`).

**Comment by stevengj** [comment]
I'm quite surprised that it worked with an escaped backslash.   If you have a file `"foo bar"`, you can't open it with `open("foo\\ bar")`.  Why would `spawn("foo\\ bar")` work?  Is this some hidden OSX or libuv feature?

**Comment by omus** [comment]
The reason the escaped backslash calls were working was due to `split_shell`

```julia
julia> subl = "/Applications/Sublime\\ Text.app/Contents/SharedSupport/bin/subl"
"/Applications/Sublime\\ Text.app/Contents/SharedSupport/bin/subl"

julia> `$subl`
`'/Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl'`

julia> run(`$subl`)
ERROR: could not spawn `'/Applications/Sublime\ Text.app/Contents/SharedSupport/bin/subl'`: no such file or directory (ENOENT)
 in _jl_spawn at process.jl:255
 in anonymous at process.jl:408
 in setup_stdio at /usr/local/Cellar/julia/HEAD/lib/julia/sys.dylib
 in __spawn#58__ at /usr/local/Cellar/julia/HEAD/lib/julia/sys.dylib
 in run at /usr/local/Cellar/julia/HEAD/lib/julia/sys.dylib

julia> import Base: shell_split; `$(shell_split(subl))`
`'/Applications/Sublime Text.app/Contents/SharedSupport/bin/subl'`

julia> run(`$(shell_split(subl))`)
# Success
```

**Comment by omus** [comment]
My goal here is to have Julia's `edit` command work with any EDITOR variable as git would. From what I can tell that means supporting:

* Editors on the path (`vim`)
* Editors including a path (`/usr/bin/vim`)
* Editors with a space in the path (`/Applications/Sublime\ Text/.../subl`, `'/Applications/Sublime Text/.../subl'`)
* Allowing flags (`subl -w`)

I may wrong here but I feel like we need to be supporting what is allowed by the shell and not forcing users to update their environment to suit Julia.

**Comment by omus** [comment]
When working within Julia with `Cmd` I agree that you should not be escaping paths since this can be handled gracefully within Julia:

```julia
julia> cmd = "foo bar"; file = "baz"; c = `$cmd $file`
`'foo bar' baz`
```

However when working with paths coming from outside of Julia, like ENV for example, I believe that you should be using `shell_split` or `shell_parse` to convert the command into something that is compatible with `Cmd`. If fact I believe that this is the purpose of the `shell_split` command. @StefanKarpinski can you comment if this is intended use for `shell_split`? I think your the creator of the function. The two other occurrences `shell_split` is used in Base Julia it appears to be used for this purpose.

**Comment by stevengj** [comment]
I can't find any sources that actually document a recommendation to backslash-escape spaces in the contents of variables like `EDITOR`.  Can you give any pointers to indicate that this is not just an idiosyncrasy of specific programs like `git`?

Even in the shell, all the sources that I can find advise using environment variables quoted like `"$THIS"` to handle embedded spaces, and most C programs will just call `getenv` to get paths and won't do any parsing to unescape backslashes.

**Comment by omus** [comment]
Unfortunately the `EDITOR` variable doesn't have that much surrounding documentation it. I'm approaching this as a bash user that is used to [quoting](http://www.gnu.org/software/bash/manual/html_node/Quoting.html) in order to avoid [word splitting](http://www.gnu.org/software/bash/manual/bash.html#index-word-splitting). Here's a quick overview of why this is necessary in bash:

```bash
$ touch '/tmp/a b'

$ bash -c "ls /tmp/a b"
ls: /tmp/a: No such file or directory
ls: b: No such file or directory

$ bash -c "ls /tmp/a\ b"
/tmp/a b

$ bash -c "ls '/tmp/a b'"
/tmp/a b
```
Julia behaves in the exact same way as bash:

```julia
julia> run(`ls /tmp/a b`)
ls: /tmp/a: No such file or directory
ls: b: No such file or directory
ERROR: failed process: Process(`ls /tmp/a b`, ProcessExited(1)) [1]
 in run at /usr/local/Cellar/julia/HEAD/lib/julia/sys.dylib

julia> run(`ls /tmp/a\ b`)
/tmp/a b

julia> run(`ls '/tmp/a b'`)
/tmp/a b

julia> run(`ls "/tmp/a b"`)
/tmp/a b

julia> string(`ls /tmp/a\ b`) == string(`ls '/tmp/a b'`) == string(`ls "/tmp/a b"`)
true
```

The fundamental reason we need to support for quoting in this particular case is to allow Julia to support an EDITOR variable that contains both spaces in a path and arguments. Otherwise "foo bar baz" means run the program `'foo bar baz'` and not run the program `foo` with arguments `bar` and `baz` (the later is bash and Julia's current behaviour).

I'm starting to get confused to what the problem is exactly. The proposed solution doesn't change the prior behaviour in regards to:

* Editors on the path (`vim`)
* Editors including a path (`/usr/bin/vim`)

But adds support for the following for all editors:

* Allowing flags (`subl -w`)
* Editors with a space in the path. Requires quoting.

I can't think of another solution to this problem that allows for both flags and spaces in paths.

**Comment by stevengj** [comment]
I see; I had thought your patch didn't support arguments, but I see that I was mistaken.

Looks reasonable to me now.  You'll need to squash the commits.

Is there any way to add a test for this?  Maybe at least some tests for `shell_split` can be added?

**Comment by omus** [comment]
I'll do some experimenting with how to test this. I definitely don't want to actually launch the editor during a test but I could probably abstract out the generation of the `Cmd` such that we can check it against what it intends to run. Does that sound reasonable?

Am I correct in that you want me to combine all of the commits before this is merged?

**Comment by stevengj** [comment]
Yes, squashing means combining all the commits.  You can do that with an interactive rebase (`git rebase -i`) followed by a forced push (`git push -f`).

**Comment by stevengj** [comment]
I think it's probably enough to add a couple of tests for `shell_split`.  Refactoring `edit` seems like more trouble than it is worth.

**Comment by omus** [comment]
I am aware of squashing but I was surprised that I was asked to do it for a PR. Note that `--force-with-lease` is safer option for a forced push.

I focused on adding tests for quoting. Let me know if you want additional tests.

**Comment by omus** [comment]
I think I can abstract out the part of `edit` that finds the editor. If I find I nice way to do it it will allow us to have reasonable test cases for `edit` that will help avoid the introduction of bugs.

**Comment by omus** [comment]
@stevengj I added test cases for `shell_split` and I went through the effort of abstracting the part of `edit` that finds the editor. The new `editor` function returns an array that can be used easily inside of backticks. This abstraction allowed me to test to ensure that both spaces and arguments are being handled correctly.

One behavioural change I did was to remove the following check:
```
ispath(editor) && !isreadable(editor) && error("can't find \"$editor\"")
```
Which isn't a very useful as it will only error if the program is exists in the system (ispath) and the permissions are not set to readable. I can add this back in if you wish.

Let me know if you are happy with all of these changes. If you are I will squash the last two changes together and this PR should be ready to be merged.

File: base/interactiveutil.jl [hunk 1]
```diff
@@ -2,7 +2,13 @@
 
 # editing files
 
-function edit(file::AbstractString, line::Integer)
+doc"""
+    editor()
+
+Determines the editor to use when running functions like `edit`. You can change the editor
+by setting JULIA_EDITOR, VISUAL, or EDITOR as an environmental variable.
```
Comment (stevengj) [inline base/interactiveutil.jl]: Document that it returns an array to splat into backticks?

**Comment by stevengj** [comment]
Looks good, though I'm not completely sure these tests should go into two separate files.

File: base/interactiveutil.jl (hunk 1 above)
Comment (omus) [inline base/interactiveutil.jl]: Will do

**Comment by omus** [comment]
I reworked the `shell_split` tests that were for edit to be more general.

> Looks good, though I'm not completely sure these tests should go into two separate files.

Do you want me to put the tests in the same file or remove the redundant tests?

**Comment by stevengj** [comment]
Maybe put the `editor` tests in `replutil.jl` and the `shell_split` tests in `spawn.jl`?

**Comment by omus** [comment]
Ok I moved the tests into the appropriate files.

**Comment by stevengj** [comment]
LGTM.  Squash the commits, then probably good to merge?

**Comment by omus** [comment]
Lets merge this! Thanks for your patience on this PR @stevengj

File: base/interactiveutil.jl [hunk 2]
```diff
@@ -2,7 +2,14 @@
...
 
-function edit(file::AbstractString, line::Integer)
+doc"""
+    editor()
+
+Determines the editor to use when running functions like `edit`. Returns an Array compatible
+for use within backticks. You can change the editor by setting JULIA_EDITOR, VISUAL, or
+EDITOR as an environmental variable.
+"""
+function editor()
```
Comment (yuyichao) [inline base/interactiveutil.jl]: Should this be sync'd to the rst doc?

Comment (omus) [inline base/interactiveutil.jl]: `editor` isn't an exported function so I would guess this should not be included in the external documentation. Opinions?

Comment (yuyichao) [inline base/interactiveutil.jl]: Should be fine to leave this as is for now....

**Comment by omus** [comment]
I randomly stumbled upon `withenv` and included it in `test/replutil.jl`

**Comment by tkelman** [comment]
My preference would be to backport this after 0.4.0, since it might require some hands-on interactive testing to get the kinks out.

**Comment by stevengj** [comment]
Yes, 0.4.1 seems fine.  I hate to see anything non-critical go into 0.4.0 this late.

**Comment by omus** [comment]
Completely fine with me.

**Pull request merged**